merge slip detection into main

Ja-r-ed opened the pull request.

ICRoboticsWorkshop committed:
added Acceleration logging, going to test what acceleration(deceleration) the robot will face when in a collision(smashing into the reef)

ICRoboticsWorkshop committed:
added is collision bool and trigger, recorded collision test against reef

ICRoboticsWorkshop committed:
fixed spelling mistake and added function with 0.6 testing value for Gs in a collision - UNTESTED

Ja-r-ed committed:
added logging in subdrivebase periodic for collision detection boolean

Ja-r-ed committed:
Added IsSlipping functions to header file

Ja-r-ed committed:
removed IsSlipping functions from header as they should be in SwerveModule not SubDrivebase

Ja-r-ed committed:
added IsSlipping functions to SwerveModule header

Ja-r-ed committed:
removed IsSlipping from swerve module header as it will need translation velocities from other swerve modules

Ja-r-ed committed:
added IsSlipping back into subdrivebase header and will use swervemodule GetSpeed to calculate slippage

Ja-r-ed committed:
added GetSlippingModule to header

Ja-r-ed committed:
need to find rotation magnitude

Ja-r-ed committed:
https://docs.wpilib.org/en/stable/docs/software/kinematics-and-odometry/swerve-drive-kinematics.html

found how to use kinematics to calculate robot speeds from module speeds

Ja-r-ed committed:
started making robot rotation and module state getting functions in subdrivebase

Ja-r-ed committed:
added swerve module state type for robot rotation function input

Ja-r-ed committed:
figuring out how ToChassisSpeeds works

Ja-r-ed committed:
finished GetRobotRotationFromStates, started on GetSlippingModule function

Ja-r-ed committed:
finished GetSlippingModule function

Ja-r-ed committed:
fixed units error in GetSlippingModule - note: dont have enough ram to build this on my laptop anymore

ICRoboticsWorkshop committed:
Merge branch 'main' into Slip+Collision-Detection

ICRoboticsWorkshop committed:
wierd issue with BL going up and then down to match others, changed to using vectors instead of just magnitudes

Gastatouillle committed:
made getslippingmodule function to return vector of slipping modules. Magically fixed issue with flatlining motor from last week. currently no slipping modules are logged

Ja-r-ed committed:
added logging of slipping booleans according to gastons idea

ICRoboticsWorkshop committed:
added a way to get the slip ratio and put getslippingmodule into drivebase periodic

ICRoboticsWorkshop committed:
created ICPoseEstimator and ICSwerveDrivePoseEstimator Utilities. TODO - make function "UpdateWithStdDevs" that updates module states with module stddevs.

ICRoboticsWorkshop committed:
removed slip ratio, now logging individual module translation values

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -87,6 +88,17 @@ SubDrivebase::SubDrivebase() {
+  
+  frc::SmartDashboard::PutNumber("Drivebase/AccelerationX", _gyro.GetAccelerationX().GetValueAsDouble());
+  frc::SmartDashboard::PutNumber("Drivebase/AccelerationY", _gyro.GetAccelerationY().GetValueAsDouble()); // smashed into reef with -1.2g and -0.4g 
+  frc::SmartDashboard::PutNumber("Drivebase/AccelerationZ", _gyro.GetAccelerationZ().GetValueAsDouble()); // normal driving goes up to around the same -1.0g
```
ideally comments are not game specific, so remove reference to reef

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -87,6 +88,17 @@ SubDrivebase::SubDrivebase() {
+  frc::SmartDashboard::PutNumber("Drivebase/AccelerationZ", _gyro.GetAccelerationZ().GetValueAsDouble()); // normal driving goes up to around the same -1.0g
+  frc::SmartDashboard::PutNumber("Drivebase/IsCollision", SubDrivebase::GetInstance().IsCollision());
+
   auto loopStart = frc::GetTime();
```
we record the loop time of each subsystem's periodic function with GetTime(). Will want to move this to the start of periodic so it records the full loop time.

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -643,4 +655,123 @@ frc2::CommandPtr SubDrivebase::WheelCharecterisationCmd() {
+
+std::vector<SubDrivebase::slippingModule> SubDrivebase::GetSlippingModule()
+{
+  // double SlipSpeed = 0.5; // in mps
```
remove if not needed

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -643,4 +655,123 @@ frc2::CommandPtr SubDrivebase::WheelCharecterisationCmd() {
+std::vector<SubDrivebase::slippingModule> SubDrivebase::GetSlippingModule()
+{
+  // double SlipSpeed = 0.5; // in mps
+
```
would be good to comment the steps in this function along the lines of our planning steps here https://github.com/icrobotics-team5584/FRC-2025-Dwight/issues/170#issuecomment-3148259389
Will help future readers understand whats going on

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -643,4 +655,123 @@ frc2::CommandPtr SubDrivebase::WheelCharecterisationCmd() {
+  Logger::Log("Drivebase/SlippingModule/blTrans", blTrans.value());
+  Logger::Log("Drivebase/SlippingModule/brTrans", brTrans.value());
+
+  // double avgTrans = (flTrans.value() + frTrans.value() + blTrans.value() + brTrans.value()) / 4.0;
```
remove if not needed

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -643,4 +655,123 @@ frc2::CommandPtr SubDrivebase::WheelCharecterisationCmd() {
+
+  std::vector<slippingModule> slippingModules;
+
+  units::meters_per_second_t _threshold = 0.5_mps;
```
local variable shouldn't start with an underscore. Only use underscores for data created at the class scope (in the header file)

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -643,4 +655,123 @@ frc2::CommandPtr SubDrivebase::WheelCharecterisationCmd() {
+          }
+      }
+  }
+  return slippingModules;
```
Do we still need this function to be retuning the slipping modules, or should its purpose just be to log slipping information?
Could simplify a lot by making this a void function and doing all the logging in the function.

Liam-Stow commented on src/main/include/subsystems/SubDrivebase.h:
```diff
@@ -85,6 +86,11 @@ class SubDrivebase : public frc2::SubsystemBase {
+  frc2::Trigger IsCollisionTrigger();
+
+  units::radians_per_second_t GetRobotRotationFromStates
+  (frc::SwerveModuleState fl, frc::SwerveModuleState fr, frc::SwerveModuleState bl, frc::SwerveModuleState br);
```
Move these out of the `// Commands` section and up into somewhere more organised, like the `// Getters` section.

Liam-Stow commented on src/main/include/subsystems/SubDrivebase.h:
```diff
@@ -178,6 +185,24 @@ class SubDrivebase : public frc2::SubsystemBase {
+    FL
+  };
+
+  std::vector<slippingModule> GetSlippingModule();
```
move to getters section or similar. I don't think the function needs to be private.

Ja-r-ed replied:
it comes up with an error saying that "slippingModule is undefined", could it be related to the enum being private?

Liam-Stow replied:
could be that `slippingModule` needs to be defined above (before)  `std::vector<slippingModule> GetSlippingModule()`. If you go down the route of turning `GetSlippingModule()` into a void function, then you could possibly just get rid of `slippingModule`

Ja-r-ed replied:
I have kept the enum but was able to move it into the getters section after turning it into a void if that is ok

Liam-Stow commented on src/main/include/subsystems/SubDrivebase.h:
```diff
@@ -178,6 +185,24 @@ class SubDrivebase : public frc2::SubsystemBase {
+  bool BLSlipping = false;
+  bool BRSlipping = false;
+
+  double SlipRatio;
```
unused variable?

Liam-Stow commented on src/main/include/subsystems/SubDrivebase.h:
```diff
@@ -178,6 +185,24 @@ class SubDrivebase : public frc2::SubsystemBase {
+  bool FLSlipping = false;
+  bool FRSlipping = false;
+  bool BLSlipping = false;
+  bool BRSlipping = false;
```
class level variables should start with underscore. If we are making GetSlippingModule() into a void func (and maybe rename to LogWheelSlipping() or similar?) then these could be removed. Might not even need the slippingModule enum.

Liam-Stow commented on src/main/include/utilities/ICPoseEstimator.h:
Do we still need the ICPoseEstimator? If not then it would be best to remove it and swap back to the wpilib one so we get any updates they make to it.

Liam-Stow reviewed (changes requested):
Added a few comments, would be good to also run the auto-formatter over your additions.

Ja-r-ed committed:
auto format GetSlippingModule()

Ja-r-ed committed:
removed "reef" in comment

Ja-r-ed committed:
moved loop time getter to top of drivebase periodic

Ja-r-ed committed:
removed unused variable "slipspeed"

Ja-r-ed committed:
commented GetSlippingModule function

Ja-r-ed committed:
moved slipping module functions from commands to getters section

Ja-r-ed committed:
removed slip ratio variable

Ja-r-ed committed:
changed slipping booleans to be underscore and lower camel case

Ja-r-ed committed:
fixed errors related to wrong variable names

Ja-r-ed committed:
moved GetSlippingModule into getters section, made GetSlippingModule into a void LogWheelSlipping

Ja-r-ed committed:
deleted IC swervedriveposeestimator and IC poseestimator, changed drivebase to use regular wpilib one instead

Liam-Stow commented:
How do you feel about removing the class-scope data (_flSlipping, _frSlipping, etc) and just doing the "SlipDetection/XXIsSlipping" logging directly inside LogWheelSlipping()? 

As a rule of thumb, it is more maintainable to keep data as local as possible (eg, if it is only used in one function, then define it within that function). When you have lots of outer-scope variables, it becomes easy to accidentally introduce bugs by modifying the data in places you didn't expect. This is the same concept as avoiding global variables where possible.

Ja-r-ed commented:
Hmmm, thinking about it more now, I realise that the slip threshold we use is completely arbitrary, I'll do some testing next week and get a better value there. I'll also log directly in the function rather than using the variables along with that next week.

ICRoboticsWorkshop committed:
updated slipping threshold, logged max and min translation values aswell as the difference between the two

ICRoboticsWorkshop committed:
removed enum and for loops, will just calculate boolean inside the log

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -14,6 +14,7 @@
 #include "RobotContainer.h"
 #include <frc/geometry/Translation2d.h>
 #include "utilities/LEDHelper.h"
+#include <map>
```
I believe we no longer need the <map> include. Removing unneeded includes will keep our compile times faster.

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -643,4 +651,94 @@ frc2::CommandPtr SubDrivebase::WheelCharecterisationCmd() {
+  return frc2::Trigger {[this] {return this->IsCollision();}};
+}
+
+
```
to be pedantic, a few extra blank lines here. One blank line between functions is the standard.

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -643,4 +651,94 @@ frc2::CommandPtr SubDrivebase::WheelCharecterisationCmd() {
+  Logger::Log("Drivebase/SlippingModule/MaxTrans", maxValue);
+
+  units::meters_per_second_t minValue = std::min({flTrans, frTrans, blTrans, brTrans});
+    Logger::Log("Drivebase/SlippingModule/MinTrans", minValue);
```
Looks like an unintentional indent?

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -643,4 +651,94 @@ frc2::CommandPtr SubDrivebase::WheelCharecterisationCmd() {
+  units::meters_per_second_t difference = maxValue - minValue;
+  Logger::Log("Drivebase/SlippingModule/biggestTransSpeedDifference", difference);
+
+  units::meters_per_second_t _threshold = 1.8_mps;
```
local variables (not class scope) shouldn't start with underscore

Liam-Stow commented on src/main/include/subsystems/SubDrivebase.h:
```diff
@@ -157,7 +165,8 @@ class SubDrivebase : public frc2::SubsystemBase {
        frc::SwerveModulePosition{0_m, _backRight.GetAngle()}},
-      frc::Pose2d()};
+      frc::Pose2d(),
+      {0.1, 0.1, 0.1}, {0.9, 0.9, 0.9}};
```
Do we want to keep these std devs or go back to letting it use a default?

Liam-Stow commented on src/main/include/subsystems/SubDrivebase.h:
```diff
@@ -67,6 +67,14 @@ class SubDrivebase : public frc2::SubsystemBase {
   frc::SwerveDriveKinematics<4> GetKinematics();
   frc::ChassisSpeeds GetRobotRelativeSpeeds();
 
+  bool IsCollision();
```
I think these should be up with the other getters, like so:

  ```cpp
// Getters
  double TranslationPosError(frc::Pose2d pose, units::meter_t startingDistance);
  double TranslationPosDistance(frc::Pose2d pose);
  bool IsAtPose(frc::Pose2d pose);
  bool IsCollision();
  frc2::Trigger CheckCoastButton();
  frc2::Trigger IsTipping();
  frc2::Trigger IsCollisionTrigger();
```

Liam-Stow commented on src/main/include/subsystems/SubDrivebase.h:
```diff
@@ -67,6 +67,14 @@ class SubDrivebase : public frc2::SubsystemBase {
+  units::radians_per_second_t GetRobotRotationFromStates
+  (frc::SwerveModuleState fl, frc::SwerveModuleState fr, frc::SwerveModuleState bl, frc::SwerveModuleState br);
+
+  void LogWheelSlipping();
```
This would make sense alongside the other instantaneous functions like so:
```cpp
 // Instantaneous functions
  void AddVisionMeasurement(frc::Pose2d pose, units::second_t timeStamp, wpi::array<double,3> dev);
  void ResetGyroHeading(units::degree_t startingAngle = 0_deg);
  ...
  void LogWheelSlipping();
```

Liam-Stow commented on src/main/include/subsystems/SubDrivebase.h:
```diff
@@ -67,6 +67,14 @@ class SubDrivebase : public frc2::SubsystemBase {
+  bool IsCollision();
+  frc2::Trigger IsCollisionTrigger();
+
+  units::radians_per_second_t GetRobotRotationFromStates
+  (frc::SwerveModuleState fl, frc::SwerveModuleState fr, frc::SwerveModuleState bl, frc::SwerveModuleState br);
```
This would make sense alongside the other kinematics related getters, like so:

```cpp
  units::meters_per_second_t GetVelocity();
  frc::SwerveDriveKinematics<4> GetKinematics();
  frc::ChassisSpeeds GetRobotRelativeSpeeds();
  units::radians_per_second_t GetRobotRotationFromStates(
    frc::SwerveModuleState fl, frc::SwerveModuleState fr, frc::SwerveModuleState bl, frc::SwerveModuleState br);
```

Liam-Stow commented on src/main/cpp/subsystems/SubDrivebase.cpp:
```diff
@@ -643,4 +651,94 @@ frc2::CommandPtr SubDrivebase::WheelCharecterisationCmd() {
+  auto GsY = _gyro.GetAccelerationY().GetValueAsDouble();
+  GsY = abs(GsY);
+  GsX = abs(GsX);
+  double Magnitude = sqrt(GsY*GsY+GsX*GsX);
```
Magnitude is a variable name, should start with lower case (magnitude).
GsY and GsX are more ok because they are literally referring to Gs of acceleration.

Liam-Stow reviewed (changes requested):
Looking good. I picked up a few more pedantic things but I think that will be all.

ICRoboticsWorkshop committed:
Removed map include
changed incorrect variables according to correct naming convention
fixed incorrect formatting
moved function declarations to correct sections in drivebase.h
